packages/next/server/web/adapter.ts

```diff
@@ -27,14 +32,40 @@ export async function adapter(params: {
   })
 
   const event = new NextFetchEvent({ request, page: params.page })
-  const original = await params.handler(request, event)
+  const response = await params.handler(request, event)
 
   return {
-    response: original || NextResponse.next(),
+    response: response || NextResponse.next(),
     waitUntil: Promise.all(event[waitUntilSymbol]),
   }
 }
 
+export function blockUnallowedResponse(
+  promise: Promise<FetchEventResult>
+): Promise<FetchEventResult> {
+  return promise.then((result) => ({
+    ...result,
+    response: isAllowed(result.response)
+      ? result.response
+      : new Response(
+          JSON.stringify({
+            message: `A middleware can not alter response's body. Learn more: https://nextjs.org/docs/messages/returning-response-body-in-middleware`,
+          }),
+          {
+            status: 500,
+            statusText: 'Internal Server Error',
+            headers: { 'content-type': 'application/json' },
+          }
+        ),
+  }))
+}
+
+function isAllowed(response: NextMiddlewareResult): boolean {
+  return (
+    !response?.body || !response.body || response.headers.has(RedirectHeader)
```

Comment:
Why is the error not shown when a custom body is added and a redirect header is present? We might want to still error for this or the user might expect the custom body to still be sent

Reply:
That's a good catch. The reason is because [the current implementation](https://github.com/vercel/next.js/blob/canary/packages/next/server/web/spec-extension/response.ts#L47-L59) of `NextResponse.redirect()` needs the ability to return a body. It's an RFC requirement: https://github.com/vercel/next.js/pull/31886

Reply:
We should be handling that internally and not from the user's response similar to the linked PR

Reply:
Recording our slack discussion:
1. change `NextResponse.redirect()` ([source](https://github.com/vercel/next.js/blob/canary/packages/next/server/web/spec-extension/response.ts#L55)) to leave the body as `null` and only set the `Location` response header.
2. rely on `generateCatchAllMiddlewareRoute()` ([source](https://github.com/vercel/next.js/blob/canary/packages/next/server/next-server.ts#L1142)) to take that header and stick it into the response body.